File: Mascot Parser Files/csharp_vs2013/example_csharp/repeat_search.cs
/*
##############################################################################
# file: repeat_search.cs                                                #
# 'msparser' toolkit example code                                            #
##############################################################################
# COPYRIGHT NOTICE                                                           #
# Copyright 1998-2015 Matrix Science Limited  All Rights Reserved.           #
#                                                                            #
##############################################################################
#    $Source: /vol/cvsroot/parser/examples/test_csharp/repeat_search.cs,v $                                #
#     $Author: patricke $                                                    #
#       $Date: 2015/07/31 14:38:19 $                                         #
#   $Revision: 1.2 $                                                         #
# $NoKeywords::                                                            $ #
##############################################################################
*/

using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using matrix_science.msparser;

namespace MsParserExamples
{
    public class repeat_search
    {
        public static void Main(string[] argv)
        {
            if (argv.Length < 1)
            {
                usage();
                return;
            }
            repeatSearch(argv[0]);
        }

        private static bool repeatSearch(string filename)
        {
            bool success = false;

            ms_mascotresfile file = new ms_mascotresfile(filename, 0, "");
            if (file.isValid())
            {
                StringBuilder s = new StringBuilder();            // Build up a MIME format string with all parameters
                s.Append("----12345\n");
                s.Append("Content-Disposition: form-data; name=\"QUE\"");
                s.Append("\n\n");

                // Parameters section
                int count = 1;
                string key = file.enumerateSectionKeys(ms_mascotresfile.section.SEC_PARAMETERS, count);
                while (key.Length != 0)
                {
                    string val = file.getSectionValueStr(ms_mascotresfile.section.SEC_PARAMETERS, key);
                    // To search against a different database, add && key != "DB"
                    if (val.Length > 0 && !key.Equals("INTERMEDIATE") && !key.Equals("RULES") && !key.Equals("INTERNALS") && !key.Equals("SEARCH"))
                    {
                        s.Append(string.Format("{0}={1}\n", key, val));
                    }
                    key = file.enumerateSectionKeys(ms_mascotresfile.section.SEC_PARAMETERS, ++count);
                }
                // To search against a different DB add: s.Append("DB=MY_DB\n");
                //s.Append("DB=repeat\n");

                // Most flexible to repeat each search as a 'sequence' search.
                s.Append("SEARCH=SQ\n");

                // For ms-ms data, tell nph-mascot where to find the ions data
                s.Append("INTERMEDIATE=").Append(filename).Append("\n");

                // Now the repeat search data
                for (int q = 1; q <= file.getNumQueries(); q++)
                {
                    s.Append(file.getRepeatSearchString(q)).Append("\n");
                }
                s.Append("----12345--\n");

                // Start nph-mascot.exe, and redirect the output to tmp.txt
                // Note that for Unix, you may need to use ./nph-mascot.exe
                try
                {
                    ProcessStartInfo start = new ProcessStartInfo();
                    start.FileName = @"./nph-mascot.exe";
                    start.Arguments = "4 -commandline";
                    start.RedirectStandardOutput = true;
                    start.RedirectStandardInput = true;
                    start.UseShellExecute = false;

                    using (Process process = Process.Start(start))
                    {
                        using (StreamWriter writer = process.StandardInput)
                        {
                            writer.Write(s.ToString());
                            writer.Flush();
                        }
                        using (StreamReader reader = process.StandardOutput)
                        {
                            string buf;
                            while ((buf = reader.ReadLine()) != null)
                            {
                                if (buf.Contains("SUCCESS"))
                                {
                                    if ((buf = reader.ReadLine()) != null)
                                    {
                                        Console.WriteLine(buf);
                                        compareResults(file, buf);
                                        success = true;
                                    }
                                }
                                else if (buf.Contains("ERROR"))
                                {

                                }
                            }
                        }
                    }

                }
                catch (Exception e)
                {
                    Console.WriteLine(e.Message);
                }
            }
            else
            {
                Console.WriteLine("Cannot open results file {0}", filename);
                Console.WriteLine(file.getLastErrorString());
            }

            return success;
        }

        private static void compareResults(ms_mascotresfile originalSearch, string repeatedSearchFileName)
        {
            ms_mascotresfile repeatedSearch = new ms_mascotresfile(repeatedSearchFileName, 0, "");
            Boolean anyReport = false;
            if (repeatedSearch.isValid())
            {
                if (originalSearch.anyPMF())
                {
                    // Use protein summary
                    ms_proteinsummary originalResults = new ms_proteinsummary(originalSearch,
                        (uint) ms_mascotresults.FLAGS.MSRES_GROUP_PROTEINS, 0, 1, null, null);
                    ms_proteinsummary repeatedResults = new ms_proteinsummary(repeatedSearch,
                        (uint)ms_mascotresults.FLAGS.MSRES_GROUP_PROTEINS, 0, 1, null, null);

                    ms_protein originalProt = originalResults.getHit(1);
                    ms_protein repeatedProt = repeatedResults.getHit(1);
                    if (originalProt != null & repeatedProt != null)
                    {
                        double diff = repeatedProt.getScore() - originalProt.getScore();
                        if (diff > 10)
                        {
                            Console.WriteLine("Protein score is {0} higher for search {1} than {2}",
                                diff, repeatedSearchFileName, originalSearch.getFileName());
                            anyReport = true;
                        }
                    }
                }
                else
                {
                    // use peptide summary
                    ms_peptidesummary originalResults = new ms_peptidesummary(originalSearch,
                        (uint)ms_mascotresults.FLAGS.MSRES_GROUP_PROTEINS, 0, 1, null, 0, 0, null);

                    ms_peptidesummary repeatedResults = new ms_peptidesummary(repeatedSearch,
                        (uint)ms_mascotresults.FLAGS.MSRES_GROUP_PROTEINS, 0, 1, null, 0, 0, null);

                    // Compare peptide scores
                    for (int q = 1; q <= originalSearch.getNumQueries(); q++)
                    {
                        ms_peptide pepOriginal = originalResults.getPeptide(q, 1);
                        ms_peptide pepRepeated = repeatedResults.getPeptide(q, 1);
                        if (pepOriginal != null && pepRepeated != null)
                        {
                            double diff = pepRepeated.getIonsScore() - pepOriginal.getIonsScore();
                            if (diff > 10)
                            {
                                Console.WriteLine("Query {0} has core {1} higher for search {2} than {3}",
                                    q, diff, repeatedSearchFileName, originalSearch.getFileName());
                                anyReport = true;
                            }
                        }
                    }
                }
                if (!anyReport)
                {
                    Console.WriteLine("Similar results for {0} and {1}", originalSearch.getFileName(), repeatedSearchFileName);
                }
            }
            else
            {
                Console.WriteLine("Invalid repeat search: {0}", repeatedSearch.getLastErrorString());
            }
        }


        private static void usage()
        {
            Console.WriteLine("Usage: repeat_search <results_file> ");
            Console.WriteLine("Given a mascot results file name, repeat the search ");
            Console.WriteLine("against the same data");
            Console.WriteLine("   result_file is a full path to a result file");
            Console.WriteLine("The program must be run from the mascot cgi directory");
        }
    }
}

/*
C:\inetpub\mascot\cgi>repeat_search.exe ..\data\F981123.dat
Will give the following output:

../data/20150728/F002065.dat
Similar results for ..\data\F981123.dat and ../data/20150728/F002065.dat
*/

File: Python/Mascot/repeatSearch.py

import msparser
import sys
import subprocess
import re

def main( filepath ) :

    return repeatSearch( filepath )


def repeatSearch(filename) :
    resfile = msparser.ms_mascotresfile(filename)

    if not resfile.isValid() :
        print("Cannot open results file %s : %s" % (filename, resfile.getLastErrorString()))
        return 1

    s = []

    s.append("----12345")
    s.append('Content-Disposition: form-data; name="QUE"')
    s.append('')

    # Parameters section
    sec_params = msparser.ms_mascotresfile.SEC_PARAMETERS
    count = 1
    key = resfile.enumerateSectionKeys(sec_params, count)

    while len(key) > 0 :
        val = resfile.getSectionValueStr(sec_params, key)

        # To search against a different database, add 'and key ne "DB"'
        if len(val) > 0 and key != "INTERMEDIATE" and key != "RULES" and key != "INTERNALS" and key != "SEARCH" :
            s.append("%s=%s" % (key, val))

        count += 1
        key = resfile.enumerateSectionKeys(sec_params, count)

    # To search against a different DB add e.g.
    # s.append("DB=MY_DB")

    # Most flexible to repeat each search as a 'sequence' search.
    s.append("SEARCH=SQ")

    # For ms-ms data, tell nph-mascot where to find the ions data
    s.append("INTERMEDIATE=" + filename)

    # Now the repeat search data.
    for q in range(1, 1 + resfile.getNumQueries()) :
        s.append(resfile.getRepeatSearchString(q))

    # Terminating line for MIME format file
    s.append("----12345--")

    # Start nph-mascot.exe, and redirect the output to tmp.txt.
    try :
        tmp = open("tmp.txt", "w")
    
    except IOError as err:
        errno, strerror = err.args
        print("Cannot open tmp.txt for writing: %s" % strerror)
        return 1
    # Replace 'nph-mascot.exe' with './nph-mascot.exe' in Unix.

    try :
        proc = subprocess.Popen(
                ['./nph-mascot.exe', '4', '-commandline'], 
                stdin=subprocess.PIPE,
                stdout=tmp
                )
    except OSError as e :
        print("Cannot run nph-mascot.exe: %s" % e)
        return 1

    for line in s :
        proc.stdin.write(line.encode('utf-8'))
        proc.stdin.write("\n".encode('utf-8'))
        

    proc.stdin.close()
    proc.wait()

    tmp.close()

    try :
        tmp = open("tmp.txt", "r")
    except IOError as err:
        errno, strerror = err.args
        print("Cannot open tmp.txt for reading: %s" % strerror)
        return 1
        
    while True :
        line = tmp.readline()

        if len(line) == 0 : break

        if re.match('.*SUCCESS.*', line) :
            # Next line contains the results file name
            line = tmp.readline()

            compareResults(resfile, line.rstrip('\n'))
            continue

        if re.match('.*ERROR.*', line) :
            print("Search failed:", line.rstrip('\n'))
            # Print details of error messages

            while len(line) != 0 :
                line = tmp.readline()
                print(line.rstrip('\n'))

            break


def compareResults(originalSearch, repeatedSearchFileName) :
    repeatedSearch = msparser.ms_mascotresfile(repeatedSearchFileName)
    anyReport = 0

    if not repeatedSearch.isValid() :
        print("Invalid repeat search: %s" % repeatedSearch.getLastErrorString())
        return

    if originalSearch.anyPMF() :
        # Use protein summary
        originalResults = msparser.ms_proteinsummary(originalSearch)
        repeatedResults = msparser.ms_proteinsummary(repeatedSearch)

        originalProt = originalResults.getHit(1)
        repeatedProt = repeatedResults.getHit(1)

        if originalProt and repeatedProt :
            diff = repeatedProt.getScore() - originalProt.getScore()

            if diff > 10.0 :
                print("Protein score is %d higher for search" % diff)
                print("%s than %s " % (originalSearch.getFileName(), repeatedSearchFileName))
                

                anyReport = 1
    else :
        #  Use peptide summary
        originalResults = msparser.ms_peptidesummary(originalSearch)
        repeatedResults = msparser.ms_peptidesummary(repeatedSearch)

        # Compare peptide scores
        for q in range(1, 1 + originalSearch.getNumQueries()) :
            pepOriginal = originalResults.getPeptide(q, 1)
            pepRepeated = repeatedResults.getPeptide(q, 1)
            diff = pepRepeated.getIonsScore() - pepOriginal.getIonsScore()

            if diff > 10.0 :
                print("Query %d has score %d higher for search %s than %s" % (q, diff, originalSearch.getFileName(), repeatedSearchFileName))
                

                anyReport = 1

    if not anyReport :
        print("Similar results for %s and %s" % (originalSearch.getFileName(), repeatedSearchFileName))
        


def usage() :
    print("""
Usage: repeat_search.py <results file>

Given an mascot results file name, repeat the search against the same data.

This program must be run in the Mascot Server cgi directory.
""")

if __name__ == "__main__" :
    sys.exit(main("data.dat"))